Added option "postProcessing".

jfendler opened the pull request:
- This option takes a comma-separated list of class names implementing
the TemplateModelPostProcessor interface.
- Before Java code is generated from templates in JavaGenerator, all
registered post-processors are called on the model, allowing them to
perform transformations on the model before Java code is created.
- Sample processors currently available are WhitespaceRemovalProcessor
and LoggingProcessor.

@jjlauer if this is more or less what you were having in mind, I hope this one is good to go. Let me know if you're not happy with the new package name ("processor"), or anything else (well, I know you will :-)

jjlauer commented:
@jfendler This is pretty awesome!  Absolutely what I had in mind.  Couple questions before I really dive into the code.  If a global processing chain is set in the maven plugin (which I think you maybe didn't include), but a user also sets the option on a per-template basis -- do those processors get added onto the end or should they replace it?

jfendler commented:
@jjlauer Thanks, glad to hear.
No, I didn't take any global configuration through maven into account. I'll look into what needs to be done for this - it'll definitely be a good addition to round this off.
But I don't really have a stong opinion on the global vs. local configuration. I can see pros and cons in both approaches. I'd probably be a little bit in favour of enforcing global configurations over per-template ones, assuming that the person editing the pom.xml is the same person as the one writing the templates and eventually doing the build. It would probably safe people a lot of useless copy-n-paste in each template.
If it turns out this might not be the best way to handle things, I could also imaging a per-template "post-processing blacklist" option, which would then allow to override the defaults as set within the global configuration. But I don't think this is the most critical thing to implement right now. I'd rather see people create some kick-ass post-processors to start with..
However, another issue to consider in this regard would be the ordering of post-processors, as there might be one PP providing something only the next one uses. Once there is support for maven global configuration of post-processors, I'd like to include a placeholder in the list to specify exactly where the local PP's would come into effect. something like
```
<postProcessing>
  my.first.Processor,{TEMPLATE_PROCESSORS},my.last.Processor
</postProcessing>

```
What are your thoughts on this?

Please do have a closer look at the code - the one thing I'm not really certain on right now (and I added this as a comment in the code as well), is whether to allow post-processors to manipulate the list of post-processors as well. I usually like fancy and flexible stuff like this, but it might lead to some tricky bugs, loops, etc. and is not addressing any particular need at this time. The existing PR code allows manipulation of all aspects of the TemplateModel (including complete replacements), but will _not_ honour any run-time changes to the list of post-processors itsel

jjlauer commented:
I think keeping it simple is the most important.

Seems like the order you list them in is the order they get applied.  Maven
plugin lets you add an ordered list.  Templates let you add an ordered
list.  If you define them in both them in both them the template list gets
appended to the maven list.   Obviously, you can't override a global
processor in this case, but a user could do any sort of fancy workaround
they need by simply creating their own processor and deciding what other
processors they simply want to call instead.  A processor can easily
instantiate its own processors (rather than needing to manipulate the
runtime list).  I don't think you'd need any sort of placeholder feature in
that case either.  If its a popular feature, I'm sure folks will submit PRs
on creating fancier processing pipelines, but let's just solve your
use-case for now.

Your PR looks pretty good as-is.  I guess adding this to the maven plugin,
fixing up the comments w/ our agreed upon approach in the above paragraph,
and then some sort of update to the documentation (probably syntax.md)
would make this a stellar PR.





On Wed, Feb 10, 2016 at 11:59 AM, Jens Fendler <notifications@github.com>
wrote:

> @jjlauer <https://github.com/jjlauer> Thanks, glad to hear.
> No, I didn't take any global configuration through maven into account.
> I'll look into what needs to be done for this - it'll definitely be a good
> addition to round this off.
> But I don't really have a stong opinion on the global vs. local
> configuration. I can see pros and cons in both approaches. I'd probably be
> a little bit in favour of enforcing global configurations over per-template
> ones, assuming that the person editing the pom.xml is the same person as
> the one writing the templates and eventually doing the build. It would
> probably safe people a lot of useless copy-n-paste in each template.
> If it turns out this might not be the best way to handle things, I could
> also imaging a per-template "post-processing blacklist" option, which would
> then allow to override the defaults as set within the global configuration.
> But I don't think this is the most critical thing to implement right now.
> I'd rather see people create some kick-ass post-processors to start with..
> However, another issue to consider in this regard would be the ordering of
> post-processors, as there might be one PP providing something only the next
> one uses. Once there is support for maven global configuration of
> post-processors, I'd like to include a placeholder in the list to specify
> exactly where the local PP's would come into effect. something like
>
> <postProcessing>
>   my.first.Processor,{TEMPLATE_PROCESSORS},my.last.Processor
> </postProcessing>
>
>
> What are your thoughts on this?
>
> Please do have a closer look at the code - the one thing I'm not really
> certain on right now (and I added this as a comment in the code as well),
> is whether to allow post-processors to manipulate the list of
> post-processors as well. I usually like fancy and flexible stuff like this,
> but it might lead to some tricky bugs, loops, etc. and is not addressing
> any particular need at this time. The existing PR code allows manipulation
> of all aspects of the TemplateModel (including complete replacements), but
> will *not* honour any run-time changes to the list of post-processors
> itsel
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fizzed/rocker/pull/21#issuecomment-182479929>.
>

jfendler commented:
@jjlauer please have a look. I have added a maven configuration option (global post-processors will be run before any per-template ones), and updated the syntax.md documentation.

jjlauer commented:
Awesome PR.  Nice job.  Will release to maven central later today.
